### PR TITLE
feat(usage): ingest live rate-limit snapshots from proxied traffic

### PR DESCRIPTION
**Pull request by Soju06**
## Why

codex-lb learns account quota state exclusively by polling `/backend-api/wham/usage` (default 60s, one account per tick), while upstream codex clients have long consumed fresher signals riding on the traffic itself. Because codex-lb discards them, selection state lags real usage by up to a refresh interval and bursty traffic can exhaust a window before the poller notices. With the 5h window temporarily removed upstream, per-turn signals are also the fastest way to observe short windows disappearing or returning.

## What

Proxied upstream responses become a passive usage source:

- **Parser + hub** (`app/core/usage/live_snapshots.py`, `live_hub.py`): `x-codex-{primary,secondary}-*` headers and `codex.rate_limits` stream events parse into typed snapshots; the client layer publishes through a startup-registered hub (no-op until the module layer registers), so core never imports modules.
- **Ingestor** (`app/modules/usage/live_ingest.py`): fire-and-forget — bounded drop-oldest queue, per-account change-fingerprint + min-interval throttle, single consumer with its own background sessions writing the same usage-history row shape the poller produces, throttled selection-cache invalidation, chatgpt-account-id resolution that drops ambiguous identities.
- **Taps**: SSE response headers at response start, stream events at the transport-agnostic `stream_responses` choke point, and `codex.rate_limits` frames in the HTTP-bridge upstream relay. All gated by a cheap marker probe; failures never touch the serving path.
- Enabled by default; `CODEX_LB_LIVE_USAGE_INGESTION_ENABLED=false` kill switch. The background poller stays authoritative for idle accounts, and its sibling-freshness gate (#1267) automatically skips accounts kept fresh by live traffic — polling pressure drops with no configuration.

## OpenSpec

New capability `live-usage-ingestion` (`openspec/changes/live-rate-limit-ingestion/`): passive snapshots, serving-path isolation, per-account throttling, decoupled hub + kill switch. Strict validation passes.

## Testing

- Unit: header/event parser edge cases, hub inertness, throttle coalescing vs prompt changed-write, queue overflow drop-oldest, `stream_responses` tap, bridge relay tap (`tests/unit/test_live_usage_ingest.py`, `test_proxy_http_bridge.py`).
- Integration: ingestor writes real usage rows for internal ids and resolved chatgpt ids; kill switch produces no publisher (`tests/integration/test_live_usage_ingest.py`).
- Full unit sweep: 3,681 passed; ty/ruff clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Soju06** [comment]
@codex review

File: app/core/clients/proxy.py [hunk 1]
```diff
@@ -2644,6 +2651,11 @@ async def _stream_via_http(
                 resp = _CodexSSEResponse(raw_resp)
                 status_code = resp.status
                 last_stream_activity_at = time.monotonic()
+                if resp.status < 400:
+                    publish_live_usage(
+                        parse_rate_limit_headers(getattr(raw_resp, "headers", None)),
+                        chatgpt_account_id=account_id,
+                    )
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/core/clients/proxy.py]: Fixed in 5f8bff9b: the direct client_session.post branch now publishes header snapshots on successful responses, same as the CodexClient route branch.

File: app/core/usage/live_snapshots.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        return float(value)
+    except ValueError:
+        return None
+
+
+def _parse_int(value: object) -> int | None:
+    parsed = _parse_float(value)
+    if parsed is None:
+        return None
+    return int(parsed)
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/core/usage/live_snapshots.py:56]: Fixed in 5f8bff9b: _parse_float rejects non-finite values (math.isfinite), so NaN/Infinity strings degrade to unparseable fields instead of raising in int() on the serving path. Covered by test_parse_rate_limit_headers_rejects_non_finite_values.

File: app/core/clients/proxy.py [hunk 3]
```diff
@@ -2510,6 +2512,11 @@ async def stream_responses(
             codex_installation_id=codex_installation_id,
             enforce_openai_sdk_contract=enforce_openai_sdk_contract,
         ):
+            if account_id and EVENT_MARKER in event_block:
+                publish_live_usage(
+                    parse_rate_limit_event_text(event_block),
+                    chatgpt_account_id=account_id,
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/core/clients/proxy.py]: Fixed in 5f8bff9b: stream_responses gains codex_lb_account_id, threaded from the streaming service mixin (via the signature-adaptive optional kwargs), quota warm-up, and limit warm-up callers; publishes prefer the internal id so multi-seat workspaces are attributed directly. Covered by the extended stream tap test.

**Comment by Soju06** [comment]
@codex review

File: app/modules/usage/live_ingest.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                    output_tokens=None,
+                    window="primary",
+                    reset_at=snapshot.primary.reset_at,
+                    window_minutes=snapshot.primary.window_minutes,
+                    credits_has=snapshot.credits_has,
+                    credits_unlimited=snapshot.credits_unlimited,
+                    credits_balance=snapshot.credits_balance,
+                )
+            if snapshot.secondary is not None:
+                await repo.add_entry(
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/usage/live_ingest.py:179]: Fixed in the latest commit: secondary rows carry the snapshot credit fields when no primary row is written (mirroring the poller convention otherwise). Covered by test_live_ingestor_carries_credits_on_secondary_only_snapshots.

**Comment by Soju06** [comment]
@codex review

File: app/modules/usage/live_ingest.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                    window_minutes=snapshot.secondary.window_minutes,
+                    credits_has=snapshot.credits_has if secondary_carries_credits else None,
+                    credits_unlimited=snapshot.credits_unlimited if secondary_carries_credits else None,
+                    credits_balance=snapshot.credits_balance if secondary_carries_credits else None,
+                )
+        self._last_write[account_id] = (_fingerprint(snapshot), time.monotonic())
+        now = time.monotonic()
+        if now - self._last_cache_invalidation >= _CACHE_INVALIDATION_MIN_INTERVAL_SECONDS:
+            self._last_cache_invalidation = now
+            get_account_selection_cache().invalidate()
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/usage/live_ingest.py]: Fixed in the latest commit: live writes invalidate get_rate_limit_headers_cache() under the same throttle as the selection cache, so downstream x-codex-* headers reflect live values before the TTL/poll fires. Covered by test_live_ingestor_invalidates_rate_limit_header_cache.

**Comment by Soju06** [comment]
@codex review

File: app/modules/quota_planner/warmup.py [hunk 6]
```diff
@@ -487,6 +487,7 @@ async def _send_warmup_probe(self, *, account: Account, model: str, request_id:
             access_token,
             upstream_account_id,
             raise_for_status=True,
+            codex_lb_account_id=account.id,
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/quota_planner/warmup.py]: Fixed in b60088bd: stream_responses gains suppress_live_usage and the quota-planner probe sets it, so the probe response cannot pre-write fresh rows and _record_warmup_effect keeps measuring via its controlled refresh. Covered by the suppression case in the stream tap test.

**Comment by Soju06** [comment]
@codex review

File: app/modules/usage/live_ingest.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                    window_minutes=snapshot.primary.window_minutes,
+                    credits_has=snapshot.credits_has,
+                    credits_unlimited=snapshot.credits_unlimited,
+                    credits_balance=snapshot.credits_balance,
+                )
+            if snapshot.secondary is not None:
+                # Mirror the poller: credits normally ride the primary row.
+                # A secondary-only snapshot (e.g. the short window is not
+                # being reported) must still carry the fresh credit state.
+                secondary_carries_credits = snapshot.primary is None
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/usage/live_ingest.py]: By design, matching the poller's semantics: when upstream stops reporting the short window the poller also leaves the last primary row in place, and the merged adaptive-windows work handles that read-side — selection state and pooled headers expire any sample whose reset_at has elapsed (#1267, expire_elapsed_window_rows + _state_from_account), and a strictly newer sibling row clears the derived short-window duration before expiry (#1268). Residual staleness is bounded by at most one short-window length and self-corrects at the stale reset, identical to poller-only operation. Making live secondary-only snapshots write primary absence markers would diverge from the poller storage contract this PR deliberately mirrors; happy to explore absence markers as a follow-up change if bounded staleness proves insufficient in practice.

File: app/core/clients/proxy.py [hunk 8]
```diff
@@ -2644,6 +2658,12 @@ async def _stream_via_http(
                 resp = _CodexSSEResponse(raw_resp)
                 status_code = resp.status
                 last_stream_activity_at = time.monotonic()
+                if resp.status < 400 and not suppress_live_usage:
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/core/clients/proxy.py]: Fixed in a847d951: headers now parse regardless of response status on both the CodexClient and direct branches, so 429/403 saturated-window snapshots are ingested; suppress_live_usage is still honored and error responses without x-codex headers remain no-ops.

**Comment by Soju06** [comment]
@codex review

File: app/modules/usage/live_ingest.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+                    window="secondary",
+                    reset_at=snapshot.secondary.reset_at,
+                    window_minutes=snapshot.secondary.window_minutes,
+                    credits_has=snapshot.credits_has if secondary_carries_credits else None,
+                    credits_unlimited=snapshot.credits_unlimited if secondary_carries_credits else None,
+                    credits_balance=snapshot.credits_balance if secondary_carries_credits else None,
+                )
+        self._last_write[account_id] = (_fingerprint(snapshot), time.monotonic())
+        now = time.monotonic()
+        if now - self._last_cache_invalidation >= _CACHE_INVALIDATION_MIN_INTERVAL_SECONDS:
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/usage/live_ingest.py]: Fixed in the latest commit: writes inside the throttle window schedule one trailing invalidation at window expiry (cancelled on stop), so every write is covered and staleness is bounded by the throttle interval (5s) rather than the header cache TTL. Covered by test_live_ingestor_trailing_invalidation_covers_throttled_writes.

**Comment by Soju06** [comment]
@codex review

File: app/modules/usage/live_ingest.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        if account_id is None:
+            return
+        if self._should_skip(account_id, item.snapshot):
+            return
+
+        snapshot = item.snapshot
+        async with get_background_session() as session:
+            repo = UsageRepository(session)
+            if snapshot.primary is not None:
+                await repo.add_entry(
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/usage/live_ingest.py:162]: Fixed in the latest commit: the updater freshness gate now also requires the account's additional usage rows (when any exist) to be within the refresh interval, so live main-window rows cannot suppress the only sync path for additional_rate_limits. Accounts without additional rows keep the live optimization. Covered by test_refresh_accounts_fetches_when_additional_usage_ages_despite_fresh_main_rows and the fresh-additional skip control.

**Comment by Soju06** [comment]
@codex review

File: app/core/usage/live_snapshots.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            payload.get("reset_at") if payload.get("reset_at") is not None else payload.get("resets_at")
+        ),
+    )
+
+
+def parse_rate_limit_event(payload: Mapping[str, Any]) -> LiveRateLimitSnapshot | None:
+    """Parse a ``codex.rate_limits`` stream event payload."""
+    if payload.get("type") != _EVENT_TYPE:
+        return None
+    rate_limits = payload.get("rate_limits")
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/core/usage/live_snapshots.py:131]: Fixed in a71ae482: parse_rate_limit_event rejects events whose limit_id/metered_limit_name/limit_name discriminator is present and not the default codex family, mirroring the header path's scoping, so model-specific buckets cannot corrupt main usage rows. Covered by test_parse_rate_limit_event_rejects_non_default_limit_families.

File: app/modules/usage/updater.py [hunk 12]
```diff
@@ -415,6 +415,18 @@ async def _refresh_account_if_stale(
...
+    async def _additional_usage_is_stale(self, account_id: str, *, now: datetime, interval_seconds: int) -> bool:
+        # The upstream fetch is the only path that syncs additional
+        # (per-model) rate limits; live main-window rows must not keep an
+        # account "fresh" while its additional rows age past the interval,
+        # or gated-model routing starves on additional_quota_data_unavailable.
+        if self._additional_usage_repo is None:
+            return False
+        latest_at = await self._additional_usage_repo.latest_recorded_at_for_account(account_id)
+        if latest_at is None:
+            return False
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/usage/updater.py]: Fixed in a71ae482: the no-row case now counts as stale, so live rows never suppress the discovery fetch — non-gated accounts keep the pre-live poll cadence (no regression vs main) and first gated-model use cannot starve. Covered by test_refresh_accounts_fetches_when_no_additional_rows_were_ever_synced.

**Comment by Soju06** [comment]
@codex review

File: app/modules/usage/live_ingest.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+                    account_id=account_id,
+                    used_percent=float(snapshot.primary.used_percent),
+                    input_tokens=None,
+                    output_tokens=None,
+                    window="primary",
```
[the remark this answers is not in the document]
Comment (Soju06) [reply app/modules/usage/live_ingest.py:167]: Fixed in the latest commit: the ingestor mirrors the poller write-time normalization — a lone primary window with the monthly duration is filed as a monthly row (with credits), never as primary. Covered by test_live_ingestor_normalizes_monthly_only_snapshots.

**Comment by Soju06** [comment]
@codex review

**Comment by Soju06** [comment]
@codex review
